File: leetcode/count-of-range-sum.py
# 区间和的个数
# https://leetcode-cn.com/classic/problems/count-of-range-sum/description/
# O(nlogn) 做法，分治、归并排序、前缀和
import math
import sys
import collections
import random
import heapq
import typing
from math import sqrt, ceil, floor, fabs, sin, cos
from typing import List, Tuple, Set, Dict, Iterable, Deque, TypeVar
from collections import deque

LOCAL, DEBUG = False, False


# LOCAL = True
# DEBUG = True


def read_strs() -> List[str]:
    return input().strip().split(' ')


def read_ints() -> List[int]:
    return list(map(int, read_strs()))


def build_string(seq: Iterable[str]) -> str:
    return ''.join(seq)


def avg2(a: int, b: int) -> int:
    return (a + b) // 2


class Solution:
    # [start, stop)
    def get_cans(self, start: int, stop: int, pre: List[int], upper: int, lower: int) -> int:
        mid = avg2(start, stop)
        ans = 0
        l = mid
        r = mid
        # [l, r)
        for i in range(start, mid):
            while l < stop and pre[l] < pre[i] + lower:
                l += 1
            while r < stop and pre[r] <= pre[i] + upper:
                r += 1
            ans += r - l
        return ans

    def merge(self, start: int, stop: int, pre: List[int]):
        temp_arr = []
        mid = avg2(start, stop)
        i = start
        j = mid
        while i < mid and j < stop:
            if pre[i] < pre[j]:
                temp_arr.append(pre[i])
                i += 1
            else:
                temp_arr.append(pre[j])
                j += 1
        if i < mid:
            while i < mid:
                temp_arr.append(pre[i])
                i += 1
        elif j < stop:
            while j < stop:
                temp_arr.append(pre[j])
                j += 1
        for offset in range(len(temp_arr)):
            pre[start + offset] = temp_arr[offset]

    # [start, stop)
    def solve(self, start: int, stop: int, pre: List[int], lower: int, upper: int) -> int:
        if stop - start == 1:
            return 0
        mid = avg2(start, stop)
        lans = self.solve(start, mid, pre, lower, upper)
        rans = self.solve(mid, stop, pre, lower, upper)
        cans = self.get_cans(start, stop, pre, upper, lower)
        self.merge(start, stop, pre)
        return lans + rans + cans

    def countRangeSum(self, nums: List[int], lower: int, upper: int) -> int:
        pre = [0 for _ in range(len(nums) + 1)]
        for i in range(1, len(nums) + 1):
            j = i - 1
            pre[i] = pre[i - 1] + nums[j]
        return self.solve(0, len(pre), pre, lower, upper)


def main():
    nums = [-2, 5, -1]
    lower = -2
    upper = 2
    print(Solution().countRangeSum(nums, lower, upper))


if __name__ == '__main__':
    sys.setrecursionlimit(int(1e5))
    if LOCAL:
        in_file = open('in.txt', 'r')
        out_file = open('out.txt', 'w')
        sys.stdin = in_file
        sys.stdout = out_file
    main()

File: leetcode/count-of-range-sum.cpp
//区间和的个数
//https://leetcode-cn.com/classic/problems/count-of-range-sum/description/
//O(n(logn)^2) 做法，分治
using namespace std;
using ll = long long;

class Solution
{
public:
	int countRangeSum(vector<int> &nums, int lower, int upper)
	{
		vector<ll> ll_nums;
		for (int num : nums)
			ll_nums.push_back(num);
		return solve(0, int(nums.size()), ll_nums, lower, upper);
	}
	static int solve(int start, int stop, const vector<ll> &nums, int lower, int upper)
	{
		if (stop - start <= 5)
			return force(start, stop, nums, lower, upper);
		int mid = start + (stop - start) / 2;
		vector<ll> left_sums = get_left_sums(nums, start, mid);
		vector<ll> right_sums = get_right_sums(nums, mid, stop);
		sort(right_sums.begin(), right_sums.end());
		int center_cnt = 0;
		for (ll lsum : left_sums)
		{
			center_cnt += get_in(right_sums, lower - lsum, upper - lsum);
		}
		int left_cnt = solve(start, mid, nums, lower, upper);
		int right_cnt = solve(mid, stop, nums, lower, upper);
		return left_cnt + right_cnt + center_cnt;
	}

	//count num in [lower, upper]
	static int get_in(const vector<ll> &arr, ll lower, ll upper)
	{
		auto p1 = lower_bound(arr.begin(), arr.end(), lower);
		auto p2 = upper_bound(arr.begin(), arr.end(), upper);
		return p2 - p1;
	}

	//[start, mid)
	static vector<ll> get_left_sums(const vector<ll> &nums, int start, int mid)
	{
		vector<ll> ret;
		ll sum = nums[mid - 1];
		ret.push_back(sum);
		for (int i = mid - 2; i >= start; i--)
		{
			sum += nums[i];
			ret.push_back(sum);
		}
		return ret;
	}

	//[mid, stop)
	static vector<ll> get_right_sums(const vector<ll> &nums, int mid, int stop)
	{
		vector<ll> ret;
		ll sum = nums[mid];
		ret.push_back(sum);
		for (int i = mid + 1; i < stop; i++)
		{
			sum += nums[i];
			ret.push_back(sum);
		}
		return ret;
	}

	static int force(int start, int stop, const vector<ll> &nums, int lower, int upper)
	{
		int len = stop - start;
		ll arr[10];
		ll pre_sum[10];
		for (int i = 1; i <= len; i++)
			arr[i] = nums[start + i - 1];
		pre_sum[0] = 0;
		for (int i = 1; i <= len; i++)
			pre_sum[i] = pre_sum[i - 1] + arr[i];
		int cnt = 0;
		//[i, j]
		for (int i = 1; i <= len; i++)
		{
			for (int j = i; j <= len; j++)
			{
				ll temp = pre_sum[j] - pre_sum[i - 1];
				if (lower <= temp && temp <= upper)
					cnt += 1;
			}
		}
		return cnt;
	}
};